SeaPkg/Library/BasePeCoffLibNegative/BasePeCoffLibNegative.c

```diff
@@ -686,6 +690,80 @@ PeCoffLoaderImageNegativeReadFromMemory (
   return RETURN_SUCCESS;
 }
 
+/**
+
+  Get 4K page aligned VMCS size.
+
+  @return 4K page aligned VMCS size
+
+**/
+UINT32
+GetAlignedVmcsSize (
+  VOID
+  )
+{
+  MSR_IA32_VMX_BASIC_REGISTER  VmxBasic;
+
+  //
+  // Read VMCS size and and align to 4KB
+  //
+  VmxBasic.Uint64 = AsmReadMsr64 (MSR_IA32_VMX_BASIC);
+  return ALIGN_VALUE (VmxBasic.Bits.VmcsSize, SIZE_4KB);
+}
+
+/**
+  Gets the MSEG base from the MSR and calculates the MSEG size.
+
+  @param[out] MsegBase  The MSEG base address.
+  @param[out] MsegSize  The MSEG size.
+
+  @retval EFI_SUCCESS    The MSEG base and size are retrieved successfully.
+  @retval EFI_NOT_FOUND  The gMsegSmRamGuid HOB is not found.
+  @retval EFI_NOT_FOUND  The PhysicalSize is 0.
+
+**/
+EFI_STATUS
+GetMsegBaseAndSize (
+  OUT EFI_PHYSICAL_ADDRESS  *MsegBase,
+  OUT UINTN                 *MsegSize
+  )
+{
+  EFI_STATUS                         Status;
+  MSR_IA32_SMM_MONITOR_CTL_REGISTER  SmmMonitorCtl;
+  CPUID_VERSION_INFO_EBX             VersionInfoEbx;
+  STM_HEADER                         *StmHeader;
+  UINTN                              NumberOfCpus;
+
+  //
+  // Find the MSEG Base address
+  //
+  SmmMonitorCtl.Uint64 = AsmReadMsr64 (MSR_IA32_SMM_MONITOR_CTL);

```

Comment:
I believe this should also check that `SmmMonitorCtl.Bits.Valid == 1`. It is supposed to be set whenever the STM image loader (UEFI) has loaded SEA into memory and set the MSEG base in the MSR. 

My understanding is that it would be set in `SmmCpuFeaturesLib` on the SMI rendezvous and it's required for the SEA VMCALL to succeed. So, it should be set during STM execution. However, `BasePeCoffNegativeLib` is also linked against the `ResponderValidationTestLib`. I'd expect it to be set there as well since MSEG should be loaded by shell and it shouldn't be cleared until reset to my understanding.